### PR TITLE
Add command for regeneration formats

### PR DESCRIPTION
**Pull request by jordygroote**
| Q | A
| --- | ---
| Bug fix? | Kinda
| New feature? | Yes
| BC breaks? | no
| Deprecations? | no
| License | MIT

#### What's in this PR?

In this PR i added a command which iterates over the current image cache and refreshes it.

You can use this command instead of `rm -rf public/media/uploads` to regenerate/refresh image cache.

#### Why?

This because removing the image cache `rm -rf public/uploads/media` is a bad idea. 
When you do this, you create a giant load on your server because it start alot of PHP-fpm processes for cutting out the images. 

We've expierenced a problem where OOMKILLER jumped in to shoot mysql and elastic and the website went down.
I was discussing this with @alexander-schranz in slack and he asked me to create a PR for this command.
https://sulu-io.slack.com/archives/C0430GGCV/p1593690512302700 

#### Example Usage

```
php bin/console sulu:media:regenerate-formats
```
Screenshot from 2020-07-06 21-59-44

### CODE REVIEW

File: src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Finder\SplFileInfo;
+
+class FormatCacheRegenerateFormats extends Command
+{
+    protected static $defaultName = 'sulu:media:regenerate:formats';
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php:27]: We decided to go in future with the following format for commands: 

`sulu:<bundle>:<command>` so for this case it should be: `sulu:media:regenerate-formats`

File: src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        $ui->writeln('DONE');
+
+        $progressBar->finish();
+
+        return 0;
+    }
+
+    private function getFileInformationArrayFromPath($path): array
+    {
+        $exploded = explode('/', $path);
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php:104]: The pattern for the route is the following `/uploads/media/<formatKey>/<segment>/<id>-<filename>` where the segment could in some configuration not exist but is not needed to call the regenrate command.

```php
$directories = explode('/', $path, 2);
$fileNameParts = explode('-', $fileName, 2);

return [
    'id' => $fileNameParts[0],
    'formatKey' => $directories[0],
    'fileName' => $exploded[1]
];
```

File: src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        $ui = new SymfonyStyle($input, $output);
+
+        $finder = new Finder();
+        $finder->in(\realpath($this->localFormatCachePath));
+        $files = $finder->files();
+
+        $progressBar = $ui->createProgressBar(\count($files));
+
+        if (!count($files)) {
+            $ui->writeln('No images to regenerate');
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php:73]: Its always good to add some information:

`sprintf('No images to regenerate found in "%s".', $this->localFormatCachePath)`

File: src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $ui = new SymfonyStyle($input, $output);
+
+        $finder = new Finder();
+        $finder->in(\realpath($this->localFormatCachePath));
+        $files = $finder->files();
+
+        $progressBar = $ui->createProgressBar(\count($files));
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php:70]: Can be initialized after `if (!count($files)) {`

File: src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        foreach ($files as $file) {
+            $fileInformation = $this->getFileInformationArrayFromPath($file->getRelativePathname());
+
+            $this->formatManager->returnImage(
+                $fileInformation['id'],
+                $fileInformation['formatKey'],
+                $fileInformation['fileName']
+            );
+
+            $progressBar->advance();
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php:90]: Like the progress output 👍

File: src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        $ui->writeln('');
+        $ui->writeln('');
+        $ui->writeln('DONE');
+
+        $progressBar->finish();
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php:97]: Got in some OS problems when outputting something before calling `finish` on the progressbar so I would move this directly after the loop.

File: src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                $fileInformation['formatKey'],
+                $fileInformation['fileName']
+            );
+
+            $progressBar->advance();
+        }
+
+        $ui->writeln('');
+        $ui->writeln('');
+        $ui->writeln('DONE');
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php:95]: Some green `success` message would be great at the end:

```php
$ui->sucess(sprintf('Finished regenerating of "%s" images.', \count($files));
```

**Comment by jordygroote** [comment]
Thank you for the feedback. I will fix these things :). Enjoy your vacation by the way!

**Comment by jordygroote** [comment]
I fixed the feedback you provided @alexander-schranz. 
BTW i'm not sure i set up this PR the correct way.
Can you check it out? :)

**Comment by nnatter** [comment]
Dont know the `MediaBundle` well enough to merge this by myself - but wanted to thank you for the contribution anyway 🙂

File: src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+    private function getFileInformationArrayFromPath($path): array
+    {
+        $exploded = \explode('/', $path);
+
+        $directories = \explode('/', $path, 2);
+        $fileNameParts = \explode('-', $directories[1], 2);
+
+        return [
+            'id' => $fileNameParts[0],
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php:110]: Didn't got time yet to test it but this looks incorrect for me.

If I interpret the code correctly you are giving the relative path to it which looks like the following:

```php
$path = '50x50/01/245-test.jpg;
```

The result should be in this case:

```php
[
     'id' => 245,
     'formatKey' => '50x50',
     'fileName' => 'test.jpg',
]
```

But for me it looks like in your case 'id' is returned as `01/245`:

Think the correct code would look like the following:

```php
$pathParts = \explode('/', $path);
$formatKey = \reset($pathParts);
$filenameParts = explode('-', \end($pathParts), 2);
$id = (int) $filenameParts[0];
$fileName = $filenameParts[1];
```

We should add a Unit test create a test directory with different cases and check there if the FormatManager is called correctly.

Test cases which should work:

 - `50x50/01/1-test.jpg` 
 - `50x50/1-test.jpg`  // no segment
 - `50x50-inset/02/1-test-test.jpg` // filename and format name with `-`

**Comment by jordygroote** [comment]
@alexander-schranz Thanks again, i think it's a good idea to create a test. However my knowledge about this is very limited.
I don't know how this should work.

I created:
`FormatCacheRegenerateFormatsTest` in `MediaBundle/Tests/Unit/Command` i'm not sure if this is correct or what needs to be in this TestClass...

**Comment by alexander-schranz** [comment]
@jordygroote As it is just a Unit test you can run phpunit from the `MediaBundle` folder the following way:

```bash
cd src/Sulu/Bundle/MediaBundle
../../../../vendor/bin/phpunit --filter="FormatCacheRegenerateFormatsTest"
```

There is also a bin/runtests script with more functionality but mostly only needed when running functional tests:

```bash
# running all tests
bin/runtests --initialize --all
bin/runtests -i -a
# running only media bundle tests
bin/runtests --initialize --target MediaBundle --no-component-tests
bin/runtests -i -t MediaBundle -C
# running only media bundle tests with phpunit filter
bin/runtests --initialize --target MediaBundle --no-component-tests --flags="--filter=FormatCacheRegenerateFormatsTest"
bin/runtests -i -t MediaBundle -C -f="--filter=FormatCacheRegenerateFormatsTest"
```

The `--initialize` will setup the database for functional tests only needed for first call.

Basically it should be something like the following (not tested):

```php

use PHPUnit\Framework\TestCase;
// ...
use Symfony\Component\Console\Application;
use Symfony\Component\Console\Tester\CommandTester;

class FormatCacheRegenerateFormats extends TestCase
{
    /**
     * @var ObjectProphecy|FormatManagerInterface
     */
    private $formatManager;

    public function setUp(): void
    {
        $this->formatManager = $this->prophesize(FormatManagerInterface::class);
    }

    public function testExecute(): void
    {
        $this->formatManager->getImage(1, '50x50', 'test.svg')
            ->shouldBeCalled();

        $this->formatManager->getImage(2, '200x', 'test.svg')
            ->shouldBeCalled();

        $this->formatManager->getImage(3, '400x400-inset', '2020-test-test.svg')
            ->shouldBeCalled();

        // create this folder with the following files (the files just need to exist don't need any content):
        // fixtures/regenerate-formats/50x50/01/1-test.svg
        // fixtures/regenerate-formats/200x/1-test.svg
        // fixtures/regenerate-formats/400x400-inset/10/1-2020-test-test.svg

        $this->executeCommand(__DIR__ . '/fixtures/regenerate-formats');
    }

    private function executeCommand(string $localFormatCachePath): void
    {
        $localFormatCachePath = __DIR__ . '/fixtures/regenerate-formats';

        $fileSystem = new Filesystem();

        $application = new Application();
            $application->add(new FormatCacheRegenerateFormats(
            $fileSystem,
            $this->formatManager->reveal(),
            $localFormatCachePath
        ));

        $command = $application->find('sulu:media:regenerate-formats');
        $commandTester = new CommandTester($this->command);

        $commandTester->execute();
    }
}
```

**Comment by jordygroote** [comment]
> @jordygroote As it is just a Unit test you can run phpunit from the `MediaBundle` folder the following way:
> 
> ```shell
> cd src/Sulu/Bundle/MediaBundle
> ../../../../vendor/bin/phpunit --filter="FormatCacheRegenerateFormatsTest"
> ```
> 
> There is also a bin/runtests script with more functionality but mostly only needed when running functional tests:
> 
> ```shell
> # running all tests
> bin/runtests --initialize --all
> bin/runtests -i -a
> # running only media bundle tests
> bin/runtests --initialize --target MediaBundle --no-component-tests
> bin/runtests -i -t MediaBundle -C
> # running only media bundle tests with phpunit filter
> bin/runtests --initialize --target MediaBundle --no-component-tests --flags="--filter=FormatCacheRegenerateFormatsTest"
> bin/runtests -i -t MediaBundle -C -f="--filter=FormatCacheRegenerateFormatsTest"
> ```
> 
> The `--initialize` will setup the database for functional tests only needed for first call.
> 
> Basically it should be something like the following (not tested):
> 
> ```
> use PHPUnit\Framework\TestCase;
> // ...
> use Symfony\Component\Console\Application;
> use Symfony\Component\Console\Tester\CommandTester;
> 
> class FormatCacheRegenerateFormats extends TestCase
> {
>     /**
>      * @var ObjectProphecy|FormatManagerInterface
>      */
>     private $formatManager;
> 
>     public function setUp(): void
>     {
>         $this->formatManager = $this->prophesize(FormatManagerInterface::class);
>     }
> 
>     public function testExecute(): void
>     {
>         $this->formatManager->getImage(1, '50x50', 'test.svg')
>             ->shouldBeCalled();
> 
>         $this->formatManager->getImage(2, '200x', 'test.svg')
>             ->shouldBeCalled();
> 
>         $this->formatManager->getImage(3, '400x400-inset', '2020-test-test.svg')
>             ->shouldBeCalled();
> 
>         // create this folder with the following files (the files just need to exist don't need any content):
>         // fixtures/regenerate-formats/50x50/01/1-test.svg
>         // fixtures/regenerate-formats/200x/2-test.svg
>         // fixtures/regenerate-formats/400x400-inset/10/1-2020-test-test.svg
> 
>         $this->executeCommand(__DIR__ . '/fixtures/regenerate-formats');
>     }
> 
>     private function executeCommand(string $localFormatCachePath): void
>     {
>         $fileSystem = new Filesystem();
> 
>         $application = new Application();
>         $application->add(new FormatCacheRegenerateFormats(
>             $fileSystem,
>             $this->formatManager->reveal(),
>             $localFormatCachePath
>         ));
> 
>         $command = $application->find('sulu:media:regenerate-formats');
>         $commandTester = new CommandTester($this->command);
> 
>         $commandTester->execute();
>     }
> }
> ```

Thanks so much for the detailed explanation. 
Impressive you coded this without testing. Was almost correct!
I added commits with the unit test, you we're right had to fix the `getFileInformationArrayFromPath` function.

It seems to work correctly now.

Let me know if i need to do more to get this PR approved :).

Screenshot from 2020-07-12 22-49-49

File: src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+        $this->fileSystem = $filesystem;
+        $this->formatManager = $formatManager;
+        $this->localFormatCachePath = $localFormatCachePath;
+    }
+
+    protected function configure()
+    {
+        $this->setDescription('Loops over sulu image cache, and regenerates the existing images')
+        ;
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Command/FormatCacheRegenerateFormats.php:57]:
```suggestion
        $this->setDescription('Loops over sulu image cache, and regenerates the existing images');
```

Not sure why php-cs didn't catch did but `;` should be in the same line. Else it looks good from miy side.

**Comment by jordygroote** [comment]
Oh wow. Typo i guess. Will change this after my holiday (till 10 august). 👍🏻

**Comment by alexander-schranz** [comment]
@jordygroote  Thank you 👍

**Pull request merged**